src/main/java/com/microsoft/azure/functions/worker/broker/JavaFunctionBroker.java

```diff
@@ -33,34 +33,40 @@ public JavaFunctionBroker(ClassLoaderProvider classLoaderProvider) {
+		FunctionDefinition functionDefinition = new FunctionDefinition(descriptor, bindings, classLoaderProvider);
+		this.methods.put(descriptor.getId(), new ImmutablePair<>(descriptor.getName(), functionDefinition));
 	}
 
 	public Optional<TypedData> invokeMethod(String id, InvocationRequest request, List<ParameterBinding> outputs)
 			throws Exception {
-		ImmutablePair<String, JavaMethodExecutor> methodEntry = this.methods.get(id);
-		JavaMethodExecutor executor = methodEntry.right;
-		if (executor == null) {
+		ExecutionContextDataSource executionContextDataSource = buildExecutionContext(id, request);
+		JavaMethodExecutor executor = FactoryJavaMethodExecutor.createJavaMethodExecutor(this.classLoaderProvider.createClassLoader());
+		executor.execute(executionContextDataSource);
+		outputs.addAll(executionContextDataSource.getDataStore().getOutputParameterBindings(true));
+		return executionContextDataSource.getDataStore().getDataTargetTypedValue(BindingDataStore.RETURN_NAME);
+	}
+
+	private ExecutionContextDataSource buildExecutionContext(String id,  InvocationRequest request)
+			throws NoSuchMethodException {
+		ImmutablePair<String, FunctionDefinition> methodEntry = this.methods.get(id);
+		FunctionDefinition functionDefinition = methodEntry.right;
+		if (functionDefinition == null) {
 			throw new NoSuchMethodException("Cannot find method with ID \"" + id + "\"");
 		}
-
 		BindingDataStore dataStore = new BindingDataStore();
-		dataStore.setBindingDefinitions(executor.getBindingDefinitions());
+		dataStore.setBindingDefinitions(functionDefinition.getBindingDefinitions());
 		dataStore.addTriggerMetadataSource(getTriggerMetadataMap(request));
 		dataStore.addParameterSources(request.getInputDataList());
-
 		ExecutionTraceContext traceContext = new ExecutionTraceContext(request.getTraceContext().getTraceParent(), request.getTraceContext().getTraceState(), request.getTraceContext().getAttributesMap());
 		ExecutionRetryContext retryContext = new ExecutionRetryContext(request.getRetryContext().getRetryCount(), request.getRetryContext().getMaxRetryCount(), request.getRetryContext().getException());
-
-		dataStore.addExecutionContextSource(request.getInvocationId(), methodEntry.left, traceContext, retryContext);
-
-		executor.execute(dataStore);
-		outputs.addAll(dataStore.getOutputParameterBindings(true));
-		return dataStore.getDataTargetTypedValue(BindingDataStore.RETURN_NAME);
+		ExecutionContextDataSource.Builder executionContextDataSourceBuilder = new ExecutionContextDataSource.Builder();
+		ExecutionContextDataSource executionContextDataSource = executionContextDataSourceBuilder
+				.invocationId(request.getInvocationId()).funcname(methodEntry.left).traceContext(traceContext)
+				.retryContext(retryContext).dataStore(dataStore).methodBindInfo(functionDefinition.getCandidate())
+				.containingClass(functionDefinition.getContainingClass()).build();
```

Comment:
(consider no builder for performance since on request hotpath)
```suggestion
		ExecutionContextDataSource executionContextDataSource = executionContextDataSourceBuilder
				.invocationId(request.getInvocationId())
				.funcname(methodEntry.left)
				.traceContext(traceContext)
				.retryContext(retryContext)
				.dataStore(dataStore)
				.methodBindInfo(functionDefinition.getCandidate())
				.containingClass(functionDefinition.getContainingClass())
				.build();
```